mtl/mtl.py

```diff
@@ -0,0 +1,268 @@
+    opnames = [None, '!', '||', '&&', '=>', 'U', 'F', 'G', 'predicate', 'bool']
+    opcodes = {'!': NOT, '&&': AND, '||' : OR, '=>': IMPLIES,
+               'U': UNTIL, 'F': EVENT, 'G': ALWAYS}
+    opstrnames = [None, 'not', 'or', 'and', 'imply', 'until', 'event', 'always',
+                  'predicate', 'bool']
+    # negation closure of operations
+    negop = (NOP, NOP, AND, OR, AND, NOP, ALWAYS, EVENT, PRED, BOOL)
+
+    @classmethod
+    def getCode(cls, text):
+        ''' Gets the code corresponding to the string representation.'''
+        return cls.opcodes.get(text, cls.NOP)
+
+    @classmethod
+    def getString(cls, op):
+        '''Gets custom string representation for each operation.'''
+        return cls.opnames[op]
+
+    @classmethod
+    def getName(cls, op):
+        '''Gets custom long string name for each operation.'''
+        return cls.opstrnames[op]
+
+class MTLFormula(object):
+    '''Abstract Syntax Tree representation of an MTL formula'''
+
+    def __init__(self, operation, **kwargs):
+        '''Constructor'''
+        self.op = operation
+
+        if self.op == Operation.BOOL:
+            self.value = kwargs['value']
+        elif self.op == Operation.PRED:
+            self.variable = kwargs['variable']
+        elif self.op in (Operation.AND, Operation.OR):
+            self.children = kwargs['children']
+        elif self.op == Operation.IMPLIES:
+            self.left = kwargs['left']
+            self.right = kwargs['right']
+        elif self.op == Operation.NOT:
+            self.child = kwargs['child']
+        elif self.op in(Operation.ALWAYS, Operation.EVENT):
+            self.low = kwargs['low']
+            self.high = kwargs['high']
+            self.child = kwargs['child']
+        elif self.op == Operation.UNTIL:
+            self.low = kwargs['low']
+            self.high = kwargs['high']
+            self.left = kwargs['left']
+            self.right = kwargs['right']
+
+        self.__string = None
+        self.__hash = None
+
+    
+    # def negate(self):
+    #     '''Computes the negation of the MTL formula by propagating the negation
+    #     towards predicates.
+    #     '''
+    #     if self.op == Operation.BOOL:
+    #         self.value = not self.value
+    #     elif self.op == Operation.PRED:
+    #         self.relation = not self.variable
+    #     elif self.op in (Operation.AND, Operation.OR):
+    #         [child.negate() for child in self.children]
+    #     elif self.op == Operation.IMPLIES:
+    #         self.right = self.right.negate()
+    #     elif self.op == Operation.NOT:
+    #         return self.child
+    #     elif self.op == Operation.UNTIL:
+    #         raise NotImplementedError
+    #     elif self.op in (Operation.ALWAYS, Operation.EVENT):
+    #         self.child = self.child.negate()
+    #     self.op = Operation.negop[self.op]
+    #     return self
+
+    # def pnf(self):
```

Comment:
The function should work for the most part. You will need to adapt the code for predicates only.